Add snowflake transient table creation support to GE init flow

jcampbell opened the pull request:
Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
Add snowflake transient table creation support to GE init flow

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

pep8speaks commented:
Thank you for submitting to Great Expectations!

* In the file [`great_expectations/cli/datasource.py`](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/great_expectations/cli/datasource.py):

> [Line 528:54](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/great_expectations/cli/datasource.py#L528): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 785:121](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/great_expectations/cli/datasource.py#L785): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 120 characters)



* In the file [`great_expectations/datasource/sqlalchemy_datasource.py`](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/great_expectations/datasource/sqlalchemy_datasource.py):

> [Line 167:121](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/great_expectations/datasource/sqlalchemy_datasource.py#L167): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 120 characters)
> [Line 180:121](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/great_expectations/datasource/sqlalchemy_datasource.py#L180): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (128 > 120 characters)



* In the file [`tests/cli/test_init_pandas.py`](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/tests/cli/test_init_pandas.py):

> [Line 442:121](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/tests/cli/test_init_pandas.py#L442): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 120 characters)





* In the file [`tests/sqlalchemy_dataset/test_sqlalchemydataset.py`](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/tests/sqlalchemy_dataset/test_sqlalchemydataset.py):

> [Line 115:38](https://github.com/great-expectations/great_expectations/blob/ade820ad00092ca5a3b38f996cf8917259db655e/tests/sqlalchemy_dataset/test_sqlalchemydataset.py#L115): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

This linter comment has been updated based on the most recent commit.

##### Comment last updated at 2020-03-11 00:20:36 UTC

jcampbell committed:
Update tests with better path guessing

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
Update tests and changelog

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell commented:
@Aylr : Please consider this ready mostly for design review and feel free to suggest a different approach. 

In this PR, I added a new prompt to `get_batch` for snowflake and bigquery; it seemed to me a good way to ensure both transparency and functionality, but please make sure it fits in your sense for the broader flow.

Aylr commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -249,7 +249,7 @@ def __init__(self, table_name=None, engine=None, connection_string=None,
             schema = None
             # dashes are special characters in most databases so use underscores
-            table_name = "ge_tmp_" + str(uuid.uuid4()).replace("-", "_")
+            table_name = "ge_tmp_" + str(uuid.uuid4())[:8]
```
nitpick: please remove the obsolete comment about underscores.

jcampbell replied:
✔️

Aylr commented on great_expectations/datasource/sqlalchemy_datasource.py:
```diff
@@ -51,16 +51,6 @@ def build_configuration(cls, data_asset_type=None, generators=None, **kwargs):
 
         """
 
-        # As of 0.9.0, we do not require generators be configured
```
❤️

Aylr commented on tests/sqlalchemy_dataset/test_sqlalchemydataset.py:
```diff
@@ -100,13 +100,14 @@ def test_only_connection_string(sa):
-        SqlAlchemyDataset('test_schema_custom', schema='example', engine=engine,
-                          custom_sql='SELECT * FROM example.fake')
-        assert "Cannot specify both schema and custom_sql." in str(err.value)
+# NOTE: 20200306: The behavior of allowing both custom_sql and schema is currently expected
```
agreed... so why not just delete all this code?

jcampbell replied:
When you say it like that....

I was viewing it a message to our future selves for when we try to really figure out how to best handle schemas. But I think your approach is better, and what it really suggests is I should do some synthesis first. I propose we add the below into a comment in the SqlAlchemyDataset class.

Several cases we need to consider:

1. Distributed backends (e.g. Snowflake and BigQuery) often use a `<database>.<schema>.<table>` syntax, but currently we are biased towards only allowing schema.table

2. In the wild, we see people using several ways to declare the schema they want to use:
  a. In the connection string, the original RFC only specifies database, but schema is supported by some backends (Snowflake) as a query parameter.
  b. As a default for a user (the equivalent of USE SCHEMA being provided at the beginning of a session)
  c. As part of individual queries.

3. We currently don't make it possible to select from a table in one query, but create a temporary table in another schema, except for with BigQuery and (now) snowflake, where you can specify the table name (and potentially trip of database, schema, table) in the batch_kwargs.

4. The SqlAlchemyDataset interface essentially predates the batch_kwargs concept and so part of what's going on, I think, is a mismatch between those. I think we should rename custom_sql -> "temp_table_query" or similar, for example.

jcampbell replied:
I added a comment to this effect in the `create_temporary_table` method.

Aylr reviewed (dismissed):
LGTM w/ a few minor nits.

jcampbell committed:
Update based on PR review

 - Add informative comment re: need to handle schema more consistently
 - Remove dead code
 - Remove stale comment

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
Merge branch 'develop' into feature/snowflake-transient

Aylr committed:
Merge branch 'develop' into feature/snowflake-transient

Aylr committed:
* fix broken test

Aylr committed:
Merge branch 'develop' into feature/snowflake-transient

Aylr committed:
* another broken tests fixed
